Integrals of summations(nested)

mijo2 opened the pull request:
#### References to other Issues or PRs

Fixes #7827 \
Closes #8771 \
Closes #8929 

#### Brief description of what is fixed or changed

--> Added the functionality of integration over summations

#### Other comments

Example:
```
>>> from sympy import *
>>> from sympy.abc import *
>>> integrate(summation(x*n, (n, 1, N)), x)
x**2*(N**2/4 + N/4)
```

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* integrals
  * Added integration over summation
<!-- END RELEASE NOTES -->

mijo2 committed:
integrals: Added piecewise sum functionality in integrals

Cherry picking commits and changes introduced in PR #8771
and #8929, the functionality is added but incorporated with
the updated codebase

mijo2 committed:
Merge branch 'master' into 7827_integration_of_sum

mijo2 committed:
integrals: Some changes for successful build

czgdp1807 commented:
Actually, I haven't worked much with integrals in SymPy. Probably, @jksuom @asmeurer can review this better. Thanks.

jksuom commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -1649,6 +1649,20 @@ def test_issue_2975():
 
+def test_issue_7827():
+    x, n, N = symbols('x n N')
+    assert integrate(summation(x*n, (n, 1, N)), x) == x**2*(N**2/4 + N/4)
```
This doesn't look correct. If `N = 2`, then we have `integrate(x + x**2, x) == x**2/2 + x**3/3` but the right hand side is `x**2*(1 + 1/2)`.

mijo2 replied:
@jksuom actually when ```N = 2``` the expression ```summation(x*n, (n, 1, N))``` becomes ```x*1 + x*2```(= ```x*3```) and integration of ```x*3``` = ```x**2 * (3/2)``` which is ```x**2*(2**2/4 + 2/4)``` = ```x**2*3/2```

jksuom replied:
Sorry, my mistake. What would be the result with `x**n`?

mijo2 replied:
@jksuom For ```x**n``` the output is:
```-N*x**2*x**N*lerchphi(x, 1, N + 2)*gamma(N + 2)/gamma(N + 3) - 2*x**2*x**N*lerchphi(x, 1, N + 2)*gamma(N + 2)/gamma(N + 3) - x - log(x - 1)```

I think the reason for this complicated output maybe because the summation takes place first and then the expanded expression is integrated. So, when we pass ```summation(x**n, (n, 1, N))``` this summation gets evaluated first and then it is integrated to give the complicated output we see above.

jksuom replied:
I think that the result could be `Sum(x**(n+1)/(n+1), (n, 1, N))`.

mijo2 replied:
Ya agreed. I would try some changes and see if I can get this result.

mijo2 replied:
@jksuom I think when we use something like ```summation```, this function expands first and then integrates. But when I tried using ```integrate(Sum(x**n, (n, 1, N)), x)``` to integrate instead, the output was ```Sum(Piecewise((x**(n + 1)/(n + 1), Ne(n, -1)), (log(x), True)), (n, 1, N))```.

asmeurer commented:
Should we be more cautious here about infinite summations? The two don't always commute.

mijo2 commented:
@asmeurer For infinite summations, if the integration summation of functions is less than infinity and summation of integration of functions is less than infinity then they are commutative. So, should we use a check if the summation of functions is infinite i.e we have infinite summations and that the summation of those functions reach infinity then we don't commute and return the result as is.

mijo2 committed:
Merge branch 'master' into 7827_integration_of_sum

mijo2 committed:
integrals: Minor changes for rebuild

Minor changes were made in this commit as the travis bot
failed unexpectedly in an unrelated test. Along with that,
the test_integrals.py  has uneven gaps between some of the
test functions and that was fixed in this commit as well.

integrals: Minor changes in the code

integrals: Added a comment for rebuild

Last time the build process stopped due to reasons not found in the
code so minor change was made for new build.

integrals: Added exception for a special case

When evaluating summation of integrals, if the integration variable
appears in the limits of a sum, then this ValueError is raised.

mijo2 commented:
Ping @asmeurer @jksuom

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -417,6 +418,15 @@ def doit(self, **hints):
 
+        if isinstance(self.function, Sum):
+            for variable in self.variables:
+                if variable in self.function.limits[0]:
```
What if sum has iterated limits?

mijo2 replied:
@jksuom I don't completely understand what sum with iterated limits is. Can you give me an example?

jksuom replied:
I was thinking of sums that have multiple indices but it seems that those have not been defined yet, so ignore this question.

But I think that these two lines could be replaced by something like

    if not any(v in self.function.limits for v in self.variables):
        ... compute the integral

Then no `else` part is needed.

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -417,6 +418,15 @@ def doit(self, **hints):
+            for variable in self.variables:
+                if variable in self.function.limits[0]:
+                    break
+                else:
```
It looks like `else:` is not necessary.

mijo2 replied:
I will have to check this part. Will get back to you regarding this.

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -419,13 +419,10 @@ def doit(self, **hints):
+                _i = self
+                _sum = self.function
+                return _sum.func(_i.func(_sum.function, *_i.limits).doit(), *_sum.limits).doit()
 
```
What should be done if an integration variable also appears in the limits of the sum, raise an exception?

mijo2 replied:
I don't think we should raise an exception. We can just output the same thing as was happening when this fix wasn't introduced, i.e output the same thing if that is ok?

jksuom replied:
Well, if an integration variable also appears as an index or a limit, there is something wrong. Shouldn't we tell the user?

mijo2 replied:
@jksuom Sorry for the late reply, had classes today and difference in time zones.

So, I don't think we should give an exception due to the following example:

```
t = integrate(Sum(x**2, (x, 1, n)), n)
```

Now, for this, ```t.function == Sum(x**2, (x, 1, n)``` and clearly, ```t.variables == [n]```, ```n in t.function.limits[0] == True``` and also if we expand the sum, or use something like ```summation```, this integral can be easily computed despite the condition.

jksuom replied:
I think that `Sum(x**2, (x, 1, n)` only makes sense for integral values of `n` though the result is a polynomial in `n`.

mijo2 replied:
@jksuom you are right. Now I see your point. If that is the case then it makes sense to raise an exception as if the integration variable appears in limits of a sum, then it makes no sense as limits of a sum is an integer(I think) and integration variable for a range is continuous. We should raise an exception.

mijo2 replied:
@jksuom I will add the exception in a format similar to below:

```
if any(v in self.function.limits[0] for v in self.variables):
    raise ValueError('Limit of the sum cannot be an integration variable.')
_i = self
_sum = self.function
return _sum.func(_i.func(_sum.function, *_i.limits).doit(), *_sum.limits).doit()
```

jksuom commented:
This looks good but one question remains: what should be done if the sum is infinite. It is possible to change the order of integration and summation if the sum of absolute values is integrable but that may be hard to verify because integration of absolute values has many issues in SymPy. Perhaps the integral should just be returned unevaluated?

mijo2 commented:
@jksuom So, I was wondering when should we leave the integration unevaluated? Like for example:

```
integrate(Sum(x**(-1*n), (n, 1, N)).doit(), x)
```

Can be integrated based on the Geometric progression formula and when ```N->oo```, then if ```mod(x) > 1``` then the above integration can be evaluated otherwise they are left unevaluated. But, what if we are just using the above code without mentioning any integration limits. Should the function evaluate this or not?

jksuom commented:
I think that it is acceptable to compute also indefinite integrals when the sum is finite.

mijo2 commented:
@jksuom So it would be alright to compute ```integrate(Sum(x**(-1*n), (n, 1, N)), x)``` right? When we will have limits we can evaluate and check if the sum goes to infinite or not and if it does, leave it unevaluated.

jksuom commented:
We can check if the limits of the sum are finite. For example, if `N.is_integer`, then also `N.is_finite` should be true.

mijo2 commented:
@jksuom But most of the times a normal user will just pass ```N``` like directly from for example ```from sympy.abc import N``` which doesn't have anything like ```N.is_finite``` or ```N.is_integer``` set. How should we handle cases like these?

jksuom commented:
I think that it is not safe to change the order of integration and summation if the limits are not finite. The user should explicitly define `N` with `integer=True`.

mijo2 commented:
@jksuom When you set N as an integer, it automatically sets ```N.is_finite == True``` is that correct way to do it?

jksuom commented:
I think that is correct.

mijo2 commented:
@jksuom I am taking an assumption that when we do something like ```integrate(Sum(f(x), (n, a, N)), x)``` then only ```a``` and ```N``` need to be integers and finite, not ```n```.

jksuom commented:
> only a and N need to be integers and finite

That seems ok to me. If they are integers, then they are automatically finite.

mijo2 committed:
integrals: Added an exception for limits of a sum of integral

When the sum which is to be integrated has non-integer or non-finite
limits, a ValueError will be raised.

integrals: Case added for handling infinite summations in an integral

When a summation is infinite or there is a non-finite term in the
expansion of a summation, the integral of summation will not be
evaluated.

integrals: Minor update for rebuild

integrals: Updated the summation of integrations part

Updated to an older solution for the fix where the failing
test case was fixed to the newer definition of integrating.

mijo2 commented:
@jksuom One of the tests case is failing, namely:

```
>>> from sympy.stats import BetaNoncentral, density, cdf
>>> from sympy import Symbol, pprint

>>> cdf(BetaNoncentral("x", 1, 1, 1), evaluate=False)(2).doit()
```

When I investigated, I realised when it tries to run the above command, somewhere it computes:

```
integrate(Sum(2**(-_k)*_x**_k*(_k + 1)*exp(-1/2)/factorial(_k), (_k, 0, oo)), _x) 
``` 

or something similar. We can do one thing is that we can try either see if: 
  1. If the limits are finite then compute the integral as is.
  2. If the limits aren't finite, compute the sum and check if the sum is finite or not with non-finite limits.

If the limits and the sum aren't finite, then we dont compute the integral or maybe raise an exception.

jksuom commented:
> If the limits are finite then compute the integral as is.

If the limits are finite, then the integrals could be computed and the sum returned.

> If the limits aren't finite, compute the sum and check if the sum is finite or not with non-finite limits.

Maybe this is what should be done, i.e. checking that the series in the integrand is convergent. This may not suffice if the series is alternating.
 
> If the limits and the sum aren't finite, then we dont compute the integral or maybe raise an exception.

In this case, the integral could be returned unevaluated. (The exception is raised only if integration variables appear in the limits.)

mijo2 commented:
@jksuom When we integrate an expression like ```integrate(Sum(x*n, (n, 1, oo)), x)``` the ideally we shouldn't evaluate this expression since I am checking ```self.function.doit().is_finite``` if that is ```False``` then we skip the evaluation otherwise proceed. But, with the above integration example, ```Sum(x*n, (n, 1, oo))``` gives ```x*oo``` as the answer and ```x*oo.is_finite``` is ```None```.

So, to tackle this problem, we can check if any of the args in ```self.function.doit()``` has a ```oo``` term in it or not. If it does, then we skip the evaluation otherwise we continue. Now, my question is should we proceed with such kind of check as this can be some more computations.

jksuom commented:
Computing an infinite sum may be expensive or even impossible. It might be cheaper to first check if the series is convergent.

mijo2 commented:
@jksuom How can I check the convergence of the series?

mijo2 commented:
@jksuom Is there another way to check the convergence for summations?

jksuom commented:
>  Is there another way to check the convergence for summations?

It seems that such a check is implemented only for series with constant coefficients. It could be used for definite integrals of series with numerical limits, but that is only a special case. It looks like integrals of infinite sums cannot currently be handled reliably. I think that the summation

    integrate(Sum(2**(-_k)*_x**_k*(_k + 1)*exp(-1/2)/factorial(_k), (_k, 0, oo)), _x) 

in the above [comment](https://github.com/sympy/sympy/pull/18491#issuecomment-583119055) should be done elsewhere before calling `integrate`.

mijo2 commented:
@jksuom So, what you are suggesting is doing the summation beforehand for sums with infinite or unknown limits. But then, won't it be similar to doing such a summation in the integral function itself to check if the summation is infinite or not. Like won't it take roughly the same computation time as it is currently taking with the latest method of evaluating the sum?

jksuom commented:
> doing the summation beforehand for sums with infinite or unknown limits

I think that the caller should be responsible for computing the sum. `Integrate` would only integrate the terms.

mijo2 commented:
@jksuom So in the code where the below line is executed:
```integrate(Sum(2**(-_k)*_x**_k*(_k + 1)*exp(-1/2)/factorial(_k), (_k, 0, oo)), _x)```

should I update it to this instead: ```integrate(Sum(2**(-_k)*_x**_k*(_k + 1)*exp(-1/2)/factorial(_k).doit(), (_k, 0, oo)), _x)```

jksuom commented:
I think that it should be

    integrate(Sum(2**(-_k)*_x**_k*(_k + 1)*exp(-S(1)/2)/factorial(_k), (_k, 0, oo)).doit(), _x)

mijo2 commented:
@jksuom Ok, so I am finally reverting back to the below:

```
        if isinstance(self.function, Sum):
            if any(not (l.is_integer and l.is_finite) for l in self.function.limits[0][1:]):
                raise ValueError(filldedent('''
                    The limits of the sum which is
                    integrated must be integer and
                    finite.'''))
            if any(v in self.function.limits[0] for v in self.variables):
                raise ValueError('Limit of the sum cannot be an integration variable.')
            _i = self
            _sum = self.function
            return _sum.func(_i.func(_sum.function, *_i.limits).doit(), *_sum.limits).doit()
```
I am not removing the commits after the above one as it might be useful later on.

jksuom commented:
That is not what I was thinking. ValueError should be raised only if the limits contain an integration variable. That could be the first test. If the sum is infinite, then it could be returned unevaluated but no error need be raised. (So there would not be the "filldedent" message.)

mijo2 commented:
Ok so I will remove the ValueError and also I will keep the limits is integer and is finite check and if the limits is infinite, then it will be unevaluated. Will that be ok?

mijo2 commented:
@jksuom Finally doing the below updated:

```
        if isinstance(self.function, Sum):
            sum_evaluate = True
            if any(not (l.is_integer and l.is_finite) for l in self.function.limits[0][1:]):
                sum_evaluate = False
            if any(v in self.function.limits[0] for v in self.variables):
                raise ValueError('Limit of the sum cannot be an integration variable.')
            if sum_evaluate:
                _i = self
                _sum = self.function
                return _sum.func(_i.func(_sum.function, *_i.limits).doit(), *_sum.limits).doit()
```

mijo2 committed:
integrals: Updated the summation part

This time ValueError will be raised only once and for infinite
Sum, the integral will be left unevaluated.

integrals: Removed sum_evaluate flag

jksuom commented:
I think that this test should come first

            if any(v in self.function.limits[0] for v in self.variables):
                raise ValueError('Limit of the sum cannot be an integration variable.')

and then

            if not all(l.is_integer and l.is_finite for l in self.function.limits[0][1:]):
                return self

I'm not sure if both `l.is_integer` and `l.is_finite` are needed. Maybe either one alone would suffice.
The sum_evaluate flag is not necessary.

mijo2 commented:
@jksuom All the changes have been made.

jksuom commented on sympy/integrals/integrals.py:
```diff
@@ -417,6 +418,17 @@ def doit(self, **hints):
+        if isinstance(self.function, Sum):
+            if any(v in self.function.limits[0] for v in self.variables):
+                raise ValueError('Limit of the sum cannot be an integration variable.')
+            if any(not l.is_finite for l in self.function.limits[0][1:]):
```
I think that `not l.is_finite` should rather be `l.is_infinite`, which is less restrictive. This would change the test with symbol `M`:
```
>>> M = Symbol('M')
>>> not M.is_finite
True
>>> M.is_infinite
>>>
```
It is likely that `M` is used to denote an integer so integration should be allowed. (Otherwise, `oo` or `-oo` would have been used.)

mijo2 committed:
integrals: Added loose restrictions on integrals of summation

l.is_finite was changed to l.is_infinite

mijo2 commented:
@jksuom I have updated ```not l.is_finite``` to ```l.is_infinite```.

jksuom commented:
I think this is ready but the commits could be squashed. Do you know how to do that?

mijo2 commented:
Yes @jksuom I will squash some commits.

jksuom commented:
There are some changes back and forth. I think that at least those should be squashed. Otherwise, I leave the decisions to you. Several logically independent commits are ok.

mijo2 commented:
@jksuom Will this be ok?

jksuom commented:
Thanks, that is ok. This is ready to be merged.
